pymc/sampling/parallel.py

```diff
@@ -96,16 +104,29 @@ def __init__(
         shared_point,
         draws: int,
         tune: int,
-        rng: np.random.Generator,
-        seed_seq: np.random.SeedSequence,
+        rng_state: RandomGeneratorState,
         blas_cores,
+        chain: int,
+        zarr_chains: list[ZarrChain] | bytes | None = None,
+        zarr_chains_is_pickled: bool = False,
     ):
-        # For some strange reason, spawn multiprocessing doesn't copy the rng
-        # seed sequence, so we have to rebuild it from scratch
-        rng = np.random.Generator(type(rng.bit_generator)(seed_seq))
+        # Because of https://github.com/numpy/numpy/issues/27727, we can't send
+        # the rng instance to the child process because pickling (copying) looses
+        # the seed sequence state information. For this reason, we send a
+        # RandomGeneratorState instead.
+        rng = random_generator_from_state(rng_state)
         self._msg_pipe = msg_pipe
         self._step_method = step_method
         self._step_method_is_pickled = step_method_is_pickled
+        self.chain = chain
+        self._zarr_recording = False
+        self._zarr_chain: ZarrChain | None = None
+        if zarr_chains_is_pickled:
+            self._zarr_chain = cloudpickle.loads(zarr_chains)[self.chain]
+        elif zarr_chains is not None:
+            self._zarr_chain = cast(list[ZarrChain], zarr_chains)[self.chain]
+        self._zarr_recording = self._zarr_chain is not None
```

Comment:
This suggests we don't have a good abstraction for backends, if we actually need to change the code to handle a new one. Can we add methods to the base class that handle this and are no-ops for the existing ones when not needed?

Reply:
I agree, we don't have a good abstraction. The `ZarrTrace` is essentially different than the `BaseTrace`. The `BaseTrace` has a single chain in mind, the `ZarrTrace` has all of them. The workers then just get views or chunks onto which they are allowed to write draws. In the other backends, the `BaseTrace` works on a single chain, handles all of the creation logic for that, and `MultiTrace` is just a thin wrapper around many chains.
In this PR, I tried to stay as close to the `BaseTrace` as possible and kept the core signatures in place.

Reply:
Do you think we could (fine if in a future PR) refactor MultiTrace to work more like ZarrTrace in regards to multiple chains so that the logic here can be shared?

Reply:
I hope so, yes. But I think that it should be done in a different PR